File: chazm-model/src/main/java/io/github/runtimemodels/chazm/entity/AttributeFactory.java
package io.github.runtimemodels.chazm.entity;

import runtimemodels.chazm.api.entity.Attribute;
import runtimemodels.chazm.api.id.UniqueId;

/**
 * The {@linkplain AttributeFactory} interface defines the APIs for constructing {@linkplain Attribute}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface AttributeFactory {

    /**
     * Constructs an {@linkplain Attribute}.
     *
     * @param id   the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
     * @param type the {@linkplain Attribute.Type} of the {@linkplain Attribute}.
     * @return an {@linkplain Attribute}.
     */
    Attribute buildAttribute(UniqueId<Attribute> id, Attribute.Type type);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/factory/AttributeFactory.kt
package runtimemodels.chazm.model.factory

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.id.AttributeId
import runtimemodels.chazm.api.id.UniqueId

/**
 * The [AttributeFactory] interface defines the APIs for constructing [Attribute]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface AttributeFactory {

    /**
     * Constructs an [Attribute].
     *
     * @param id   the [UniqueId] that represents the [Attribute].
     * @param type the [Attribute.Type] of the [Attribute].
     * @return an [Attribute].
     */
    fun buildAttribute(id: AttributeId, type: Attribute.Type): Attribute

}